module/dummy_psu.c

```diff
@@ -135,58 +519,73 @@ long dummy_psu_ioctl(struct file *filep, unsigned int cmd, unsigned long arg)
 				       pobj->dev_name);
 				return -EFAULT;
 			}
-			pr_debug("%s: Adding power_supply_property %d",
-				 pobj->dev_name, psp);
-			pobj->props[pobj->num_properties] = psp;
+			if (copy_from_user(&psp, (void __user *)arg,
+					   sizeof(char[MAX_KEYLENGTH])) != 0)
+				return -EFAULT;
+			parsed_psp = strtoenum(psp);
+			if ((parsed_psp >= MAX_PSU_PROPS) || (parsed_psp < 0)) {
+				pr_err("IOCTL_PSU_ADD_PSP: %d: Invalid property.\n",
+				       parsed_psp);
+				return -EFAULT;
+			}
+			pr_debug(
+				"IOCTL_PSU_ADD_PSP: %s: Adding power_supply_property %s (%d)",
+				pobj->dev_name, psp, parsed_psp);
+			pobj->props[pobj->num_properties] = parsed_psp;
 			pobj->num_properties += 1;
 		} else {
-			pr_err("Cannot find PSU specs. Missing IOCTL_PSU_CREATE call?\n");
+			pr_err("IOCTL_PSU_ADD_PSP: Cannot find PSU specs. Missing IOCTL_PSU_CREATE call?\n");
 			return -EINVAL;
 		}
 		break;
 	case IOCTL_PSU_ADD_USB_TYPE:
 		if (filep->private_data) {
 			pobj = filep->private_data;
-			if (copy_from_user(
-				    &usbtype, (void __user *)arg,
-				    sizeof(enum power_supply_usb_type)) != 0)
-				return -EFAULT;
 			if (pobj->num_usb_types >= MAX_PSU_UBS_TYPES) {
-				pr_err("%s: Cannot add USB type. Number of USB types is reached\n",
+				pr_err("IOCTL_PSU_ADD_USB_TYPE: %s: Cannot add USB type. Number of USB types is reached\n",
 				       pobj->dev_name);
 				return -EFAULT;
 			}
-			pobj->usb_types[pobj->num_usb_types] = usbtype;
+			if (copy_from_user(&usbtype, (void __user *)arg,
+					   sizeof(char[MAX_KEYLENGTH])) != 0)
+				return -EFAULT;
+			parsed_usbtype = strtoenum(usbtype);
+			if ((parsed_usbtype >= MAX_PSU_UBS_TYPES) ||
+			    (parsed_usbtype < 0)) {
+				pr_err("IOCTL_PSU_ADD_USB_TYPE: %d: Invalid USB type.\n",
+				       parsed_usbtype);
+				return -EFAULT;
+			}
+			pr_debug(
+				"IOCTL_PSU_ADD_USB_TYPE: %s: Adding USB type %d",
+				pobj->dev_name, parsed_usbtype);
+			pobj->usb_types[pobj->num_usb_types] = parsed_usbtype;
 			pobj->num_usb_types += 1;
 		} else {
-			pr_err("Cannot find PSU specs. Missing IOCTL_PSU_CREATE call?\n");
+			pr_err("IOCTL_PSU_ADD_USB_TYPE: Cannot find PSU specs. Missing IOCTL_PSU_CREATE call?\n");
 			return -EINVAL;
 		}
 		break;
 	case IOCTL_PSU_ADD_SUPPLIED_TO:
 		if (filep->private_data) {
 			pobj = filep->private_data;
-			if (copy_from_user(&supplied_to, (void __user *)arg,
-					   sizeof(char[MAX_KEYLENGTH])) != 0)
-				return -EFAULT;
 			if (pobj->num_supplicants >= MAX_SUPPLICANTS) {
-				pr_err("%s: Cannot add supplicant. Number of supplicants is reached\n",
+				pr_err("IOCTL_PSU_ADD_SUPPLIED_TO: %s: Cannot add supplicant. Number of supplicants is reached\n",
 				       pobj->dev_name);
 				return -EFAULT;
 			}
+			if (copy_from_user(&supplied_to, (void __user *)arg,
+					   sizeof(char[MAX_KEYLENGTH])) != 0)
+				return -EFAULT;
 			pobj->supplied_to[pobj->num_supplicants] = supplied_to;
```

Comment:
`supplied_to` is a local variable, you can't  use it in long-lived struct. You need strdup or similar, but make sure it is 0-terminated first.